src/detect-engine-build.c

```diff
@@ -1126,9 +1126,188 @@ static int RuleSetWhitelist(Signature *s)
     return wl;
 }
 
-int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, DetectPort **newhead,
-        uint32_t unique_groups, int (*CompareFunc)(DetectPort *, DetectPort *));
-int CreateGroupedPortListCmpCnt(DetectPort *a, DetectPort *b);
+static int CreateGroupedPortListCmpCnt(const DetectPort *a, const DetectPort *b)
+{
+    if (a->sh->init->score < b->sh->init->score) {
+        return 1;
+    } else if (a->sh->init->score > b->sh->init->score) {
+        return -1;
+    }
+
+    if (a->sh->init->sig_cnt < b->sh->init->sig_cnt) {
+        return 1;
+    } else if (a->sh->init->sig_cnt > b->sh->init->sig_cnt) {
+        return -1;
+    }
+    if (a->sh->id < b->sh->id) {
+        return -1;
+    } else {
+        return 1;
+    }
+}
+
+static int SortCompare(const void *a, const void *b)
+{
+    const DetectPort *pa = *(const DetectPort **)a;
+    const DetectPort *pb = *(const DetectPort **)b;
+
+    return CreateGroupedPortListCmpCnt(pa, pb);
+}
+
+static inline void SortGroupList(uint32_t *groups, DetectPort **list,
+        int (*CompareFunc)(const DetectPort *, const DetectPort *))
+{
+    int cnt = 0;
+    for (DetectPort *x = *list; x != NULL; x = x->next) {
+        BUG_ON(x->port > x->port2);
+        cnt++;
+    }
+    if (cnt <= 1)
+        return;
+
+    DetectPort **array = (DetectPort **)SCCalloc(cnt, sizeof(DetectPort *));
+    if (array == NULL)
+        return;
+
```

Comment:
comment: build temp array to sort with qsort. Assign a temporary id to resolve otherwise equal groups.